addons/binding/org.openhab.binding.mihome/src/main/java/org/openhab/binding/mihome/internal/api/manager/MiHomeApiManagerImpl.java

```diff
@@ -0,0 +1,204 @@
+ *
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+package org.openhab.binding.mihome.internal.api.manager;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.Base64;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Properties;
+
+import org.apache.commons.io.IOUtils;
+import org.eclipse.jetty.client.api.ContentResponse;
+import org.eclipse.jetty.http.HttpStatus;
+import org.openhab.binding.mihome.internal.api.JSONResponseHandler;
+import org.openhab.binding.mihome.internal.api.constants.DeviceConstants;
+import org.openhab.binding.mihome.internal.api.constants.DeviceTypesConstants;
+import org.openhab.binding.mihome.internal.rest.RestClient;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.gson.Gson;
+import com.google.gson.JsonObject;
+import com.google.gson.JsonParseException;
+
+/**
+ * A helper class for executing requests to the Mi|Home REST API.
+ * For more information see {@link https://mihome4u.co.uk/docs/api-documentation}
+ *
+ * @author Svilen Valkanov - Initial contribution
+ * @author Mihaela Memova - Added {@link FailingRequestHandler}
+ */
+public class MiHomeApiManagerImpl implements MiHomeApiManager {
+
+    // MiHome controllers IDs
+    public static final String CONTROLLER_DEVICES = "devices";
+    public static final String CONTROLLER_SUBDEVICES = "subdevices";
+    public static final String CONTROLLER_USERS = "users";
+
+    // Actions
+    public static final String ACTION_CREATE = "create";
+    public static final String ACTION_LIST = "list";
+    public static final String ACTION_DELETE = "delete";
+    public static final String ACTION_SHOW = "show";
+    public static final String ACTION_UPDATE = "update";
+    /**
+     * This is only internal Mi|Home firmware upgrade. There is no information about the device firmware description.
+     */
+    public static final String ACTION_UPDATE_FIRMWARE = "update_latest_firmware";
+
+    private MiHomeApiConfiguration configuration;
+    private RestClient restClient;
+    private FailingRequestHandler failingRequestHandler;
+
+    private Logger logger = LoggerFactory.getLogger(this.getClass().getName());
+
+    public MiHomeApiManagerImpl(MiHomeApiConfiguration configuration, RestClient restClient,
+            FailingRequestHandler requestHandler) {
+        this.configuration = configuration;
+        this.restClient = restClient;
+        this.failingRequestHandler = requestHandler;
+    }
+
+    @Override
+    public JsonObject registerGateway(String label, String authCode) {
+        Map<String, Object> parameters = new HashMap<String, Object>();
+        parameters.put(DeviceConstants.DEVICE_TYPE_KEY, DeviceTypesConstants.GATEWAY_TYPE);
```

Comment:
I would use real objects to pass to JSON then everything becomes typed and you the option to refactor things properly.